lib/coverband/adapters/hash_redis_store.rb

```diff
@@ -0,0 +1,164 @@
+# frozen_string_literal: true
+
+module Coverband
+  module Adapters
+    class HashRedisStore < Base
+      FILE_KEY = 'file'
+      FILE_LENGTH_KEY = 'file_length'
+      META_DATA_KEYS = [DATA_KEY, FIRST_UPDATED_KEY, LAST_UPDATED_KEY].freeze
+      ###
+      # This key isn't related to the coverband version, but to the interal format
+      # used to store data to redis. It is changed only when breaking changes to our
+      # redis format are required.
+      ###
+      REDIS_STORAGE_FORMAT_VERSION = 'coverband_3_3'
+
+      attr_reader :redis_namespace
+
+      def initialize(redis, opts = {})
+        super()
+        @redis_namespace = opts[:redis_namespace]
+        @format_version = REDIS_STORAGE_FORMAT_VERSION
+        @redis = redis
+        @ttl = opts[:ttl] || -1
+      end
+
+      def clear!
+        old_type = type
+        Coverband::TYPES.each do |type|
+          self.type = type
+          file_keys = files_set
+          @redis.del(*file_keys) if file_keys.any?
+          @redis.del(files_key)
+        end
+        self.type = old_type
+      end
+
+      def clear_file!(file)
+        roots = Coverband.configuration.all_root_paths
+        absolute_path = relative_path_to_full(file, roots)
+        file_hash = file_hash(absolute_path)
+        relative_path_file = full_path_to_relative(file)
+        Coverband::TYPES.each do |type|
+          @redis.del(key(relative_path_file, type, file_hash: file_hash))
+        end
+        @redis.srem(files_key, relative_path_file)
+      end
+
+      def save_report(report)
+        roots = Coverband.configuration.all_root_paths
+        report_time = Time.now.to_i
+        updated_time = type == Coverband::EAGER_TYPE ? nil : report_time
+        script_id = hash_incr_script
+        @redis.pipelined do
+          keys = report.map do |file, data|
+            relative_file = full_path_to_relative(file)
+            absolute_path = relative_path_to_full(file, roots)
```

Comment:
@danmayer - There's some refactoring i need to do with calculating file_hash. Feel like this conversion back and forth between relative and absolute files over and over is a little messy and probably not great for performance. Thinking about pushing this down to the file_hash method and caching the hash by relative file to avoid converting from relative to absolute file each time. Some work still to do there.

Reply:
yeah, good call I think if we worked on our reporter we could try to have everything work as a relative path and avoid all the back and forth. We still have to always convert the Coverage stdlib to relative path though.